### ISSUE TITLE
[Core] Support for Faceting / Aggregations

### ISSUE THREAD

**Issue by alexander-schranz** [schranz-search/schranz-search#182]
Typical shops will allow if you search something that they show categories to add additional filters. In that case mostly only categories are shown which matches the current search condition of the user. This kind of things are called aggregations or faceting.

There are different kind of aggregations which would be nice to support but requires a lot of research if all kind of currently supported search engines support such kind of aggregations:

 - category filter aggregations (matching categories with search term)
 - price range filter aggregations (lowest and highest price)
 
###  What kind of search engines support this kind of aggregations
  
Supported:

 - Elasticsearch / Opensearch (https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html @wachterjohannes maybe you can show how currently this was done in such projects). Interesting can also be post filters: https://www.elastic.co/guide/en/elasticsearch/reference/7.17/filter-search-results.html#post-filter 

Maybe:
 
 - RediSearch (not tested yet (https://redis.io/docs/stack/search/reference/aggregations/)
 - Algolia (not tested yet (https://www.algolia.com/doc/guides/managing-results/refine-results/faceting/#retrieving-facets))
 - Meilisearch (not tested yet (https://www.meilisearch.com/docs/learn/advanced/faceted_search#facet-distribution))
 - Solr (not tested yet (https://solr.apache.org/guide/8_8/search-sample.html#facet))
 - Typesense ( not tested yet (https://typesense.org/docs/28.0/api/search.html#facet-results))

 
### Should aggregations be part of the search result or own call?

Some search engines support aggregations as combination with the search query, some already requires that the aggregation is a seperate query. Not sure what from the DX point if `SearchBuilder` allows to query aggregations or if a seperate `AggregationBuilder` would be better and provide easier DX.

### Which type of aggregations are required?

 - Term Aggregation? (e.g.: category filter)
 - Min/Max Aggregation? (e.g.: price range filter)
 - Average Aggregation (show average price value)
 - ...? What other aggregations exists?

**Comment by alexander-schranz** [comment]
@Toflar do you have anything planned to support aggregations for @loupe-php?

**Comment by Toflar** [comment]
I have no plans for nothing :) Loupe will be what people make it :)

**Comment by alexander-schranz** [comment]
I'm little bit unhappy with the terms facets and aggregations. Maybe it is because I'm not a native english speaker but I just found that terms never self explaining or deep technically.

So instead of something like `addFacet` or `addAggregation` (see https://github.com/PHP-CMSIG/search/pull/498) something like `addInsight` / `addInsightQuery` would be more self explaining.

```php
$result = $this->engine->createSearchBuilder('blog')
    ->addFilter(/* ... */)
    ->addInsightQuery(new Insight\CountInsight(field: 'category'))
    ->addInsightQuery(new  Insight\MinMaxInsight(field: 'price'))
    ->getResult();
```

But not sure 🤔 

@Toflar what do you think?

**Comment by Toflar** [comment]
Depends on what you want to do? I have a very clear definition in mind of what an aggregation and what a facet is😅

**Comment by alexander-schranz** [comment]
What definition do you have in mind for aggregation / facet?

---

For me the most common usecases which make sense for SEAL are.

### A. The "aggregated" count of matches in specific product categories, like here:

<img width="247" alt="Image" src="https://github.com/user-attachments/assets/83ffd81f-c57e-42d4-9d20-18e1d9c16a78" />


### B. A Min / Max values for a Price or other Value slider, similar to Amazon:

<img width="252" alt="Image" src="https://github.com/user-attachments/assets/5becfb35-9976-4ce0-aa07-6a0684c80a7a" />

---

In Meilisearch A is returned as facetDistribution and B as facetStats, still api uses in both cases facets.  

Stats or Metrics is something which maybe are more beginner friendly terms.

---

Think the result in best case depends on the type a little bit, not really sure if `AVG` / `SUM` are used in common usages but are theoretical aggreations which could be kept in mind:

 - `INT/FLOAT` fields: `MIN`, `MAX`, `AVG`, `SUM`, `COUNT_PER_ITEM` (Distribution)
 - `DateTime` fields: `MIN`, `MAX`, `AVG`, `COUNT_PER_ITEM` (Distribution)
 - `Bool` fields: `COUNT_PER_ITEM`
 - `Text` fields: `COUNT_PER_ITEM`

So maybe something like this could be the result not sure if I want to diff between distribution and statis like Meilisearch does ([here](https://www.meilisearch.com/docs/learn/filtering_and_sorting/search_with_facet_filters#use-facets-in-a-search-query)):

```js
{
    // numeric based fields (INT, FLOAT)
    "rating": {
        "count_per_item": {
             "2.5": 2,
             "3.0": 1,
             "4.0": 2
        },
        "avg": 2.8,
        "sum": 14.0,
        "min": 2.5,
        "max": 4,
    },
    // date based fields
    "createdAt": {
        "count_per_item": {
             "2023-01-24": 2,
             "2024-01-24": 1,
             "2025-01-24": 2
        },
        "avg": "2024-01-23",
        "min": "2023-01-24",
        "max": "2025-01-24",
    },
    // bool fields
    "isSpecial": {
        "count_per_item": {
             "true": 1,
             "false": 4
        }
    },
    // Text based fields
    "tag": {
        "count_per_item": {
             "UI": 2,
             "UX": 1,
             "Tech": 2
        }
    }
}
```

While `count_per_item` looks strange for date or numberic values. Airbnb shown that it can be kind of interesting also for that cases:

<img width="532" alt="Image" src="https://github.com/user-attachments/assets/b8f1b51f-039a-4547-9c7a-5b455a390854" />

**Comment by Toflar** [comment]
I'm not really sure either one of them can be abstracted. The faceting maybe. But the aggregation features for analytics/statistics do not seem like a regular search engine requirement to me. If you compare SEAL to Doctrine DBAL, this sounds like a more complex, driver-specific query to me...

**Comment by alexander-schranz** [comment]
Yeah what is possible in the abstraction I will then try out. Typical thing like this:

<img width="247" alt="Image" src="https://github.com/user-attachments/assets/83ffd81f-c57e-42d4-9d20-18e1d9c16a78" />

is what I think in the first review (beside Loupe) is supported by the engines. But need to test it in deep because I know from the past that it behaves differently if you already have a "filter" on one category. Some engines support to say there filter are `AND` or `OR` connected. So sure a deeper research is required here.

Time to jump into the rabbit hole:

![Rabbit Looks into the hole and then jump into it](https://github.com/user-attachments/assets/a2082f7e-739d-4cd0-98c0-9e563f07be34)

**Comment by Toflar** [comment]
Facets [are landing soon](https://github.com/loupe-php/loupe/pull/184) in Loupe... 🥳

**Comment by alexander-schranz** [comment]
@Toflar thats awesome :)
